### PR TITLE
Update (mostly codegen) to compile with nightly Rust.

### PR DESCRIPTION
**Pull request by birkenfeld**
This is unfortunately duplicating a lot from #38 by @webmobster, which I saw only after doing the work.

However, I solved some things a little differently, so I want to offer the PR anyway. Notably, I didn't have to introduce special handling for the lifetime in `DiagnosticBuilder`, by introducing a lifetime to the Tokenizer trait. (Maybe this is also due to a change in nightly between the PRs?)

To check everything builds, I used the codegen/tests crate, with `cargo test` and `cargo test --features=syntex`. Checked with syntex 0.26.

### CODE REVIEW

**Comment by webmobster** [comment]
This looks a lot cleaner than my pull, I'll close my one in favour of this one.

**Comment by Binero** [comment]
What's stopping this from merging?

File: codegen/src/parser.rs [hunk 1]
```diff
@@ -16,25 +16,26 @@ use syntax::codemap::Span;
...
-use syntax::diagnostic::FatalError;
 use syntax::parse::parser::Parser;
+use syntax::errors::DiagnosticBuilder;
 use syntax::ptr::P;
 
-trait Tokenizer {
+
+trait Tokenizer<'a> {
     // returns the current token, without consuming it
-    fn token<'a>(&'a self) -> &'a token::Token;
```
Comment (LeoTestard) [inline codegen/src/parser.rs]: I don't like lifetime ellision rules but I guess I'll accept this one...

File: codegen/tests/src/test.in.rs [hunk 2]
```diff
@@ -10,9 +10,11 @@ pub enum Token {
 
 rustlex! SimpleLexer {
     let A = 'a';
+    . => |_:&mut SimpleLexer<R>|     None
```
Comment (LeoTestard) [inline codegen/tests/src/test.in.rs]: Why this change?

**Comment by LeoTestard** [comment]
Merging this PR.
I'm really sorry about the very long delay. :( I had plenty of time to do this, I have no excuse. I'll try to be more reactive in the future.

**Comment by LeoTestard** [comment]
Merging this PR.
I'm really sorry about the very long delay. :( I had plenty of time to do this, I have no excuse. I'll try to be more reactive in the future.

**Comment by LeoTestard** [comment]
Merging this PR.
I'm really sorry about the very long delay. :( I had plenty of time to do this, I have no excuse. I'll try to be more reactive in the future.

**Comment by LeoTestard** [comment]
Merging this PR.
I'm really sorry about the very long delay. :( I had plenty of time to do this, I have no excuse. I'll try to be more reactive in the future.
